bhttp/src/http_compat/encode.rs

```diff
@@ -0,0 +1,441 @@
+                            write_vec(&data, &mut chunk_data)?;
+                            Poll::Ready(Some(Ok(chunk_data)))
+                        } else if let Some(trailers) = frame.trailers_ref() {
+                            // Is a trailers frame
+                            let mut buf = Vec::new();
+                            // First, we need to write zero-length chunk to indicate end, since we always using Mode::IndeterminateLength
+                            write_len(0, &mut buf)?;
+
+                            // Then, write the trailer field section with indeterminate length mode
+                            let field_section = FieldSection::try_from(trailers)?;
+                            field_section.write_bhttp(Mode::IndeterminateLength, &mut buf)?;
+
+                            // Switch to Done state since we have written the trailers
+                            this.state = HttpEncodeState::Done;
+                            Poll::Ready(Some(Ok(buf)))
+                        } else {
+                            // For unknown frame types, report an invalid data error.
+                            this.state = HttpEncodeState::Done;
+                            Poll::Ready(Some(Err(Error::UnknownHttpBodyFrameType)))
+                        }
+                    }
+                    Poll::Ready(Some(Err(e))) => Poll::Ready(Some(Err(Error::Io(
+                        std::io::Error::new(std::io::ErrorKind::Other, e),
+                    )))),
+                    Poll::Ready(None) => {
+                        // End of body, write zero-length chunk to indicate end
+                        let mut end_chunk = Vec::new();
+                        write_len(0, &mut end_chunk)?;
+                        this.state = HttpEncodeState::Done;
+                        Poll::Ready(Some(Ok(end_chunk)))
+                    }
+                    Poll::Pending => Poll::Pending,
+                }
+            }
+            HttpEncodeState::Done => Poll::Ready(None),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use bytes::Bytes;
+    use futures::executor::block_on;
+    use futures::StreamExt;
+    use http::{Request, Response};
+    use http_body::Frame;
+    use http_body_util::Full;
+    use http_body_util::StreamBody;
+
+    #[test]
+    fn test_request_to_bhttp_conversion() {
+        let request = Request::builder()
+            .method("GET")
+            .uri("https://example.com/path")
+            .header("User-Agent", "test-agent")
+            .header("Accept", "application/json")
+            .body(Full::new(Bytes::from("test body")))
+            .unwrap();
+
+        let stream = BhttpEncoder::from_request(request);
+        let future = stream.fold(Vec::new(), |mut acc, item| async move {
+            if let Ok(data) = item {
+                acc.extend(data);
+            }
+            acc
+        });
+
+        let bhttp_data = block_on(future);
```

Comment:
This seems like an extractable test utility.

Reply:
I just realized we can modify this

```rust
        let stream = BhttpEncoder::from_request(request);
        let future = stream.fold(Vec::new(), |mut acc, item| async move {
            if let Ok(data) = item {
                acc.extend(data);
            }
            acc
        });

        let bhttp_data = block_on(future);
```
to this
```rust
        let mut reader = BhttpEncoder::from_request(request)
            .map_err(std::io::Error::other)
            .into_async_read();
        let bhttp_data = reader.sync_read_to_end();
```

Would we still need to perform the extraction?